### PR TITLE
Improve support for autowiring of entity repositories

### PR DESCRIPTION
**Pull request by hlohaus**
### 1. Why is this change necessary?
Improve autowiring

### 2. What does this change do, exactly?
It adds a new compiler pass for the container.

### 4. Please link to the relevant issues (if any).
I can't create new, because they are 3 tracking scrip where throwing a exception.

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/workflow/2020-08-03-implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

**Comment by keulinho** [comment]
Thx for your contribution :blue_heart: :tada: 

But the issue should already be fixed by a recent commit: https://github.com/shopware/platform/commit/86dff2dd51303bca72f7b6bc196871226c7f9a33

If that fix should not work for you feel free to reopen your PR.

**Comment by hlohaus** [comment]
@keulinho No. SaleChannel Repo and Difinition missing... @keulinho 

Not only your bug fix..

**Comment by keulinho** [comment]
Ah i see, please adjust the `SalesChannelEntityCompilerPass` for the missing argument aliases for the SalesChannelRepos, the same way it is done for the "normal" repos.

Whats the need for the entity definitions? As they all have a concrete class you can autowire them with `ProductDefinition $nameDoesNotMatter` which is IMHO more stable and does not rely on variable naming, and i see no upside in type hinting only the `EntityDefinition` base class.

**Comment by hlohaus** [comment]
@keulinho thank you

but if i reuse the core compiler class, i can't wrote unit tests..?

**Comment by keulinho** [comment]
Why not? Take a look at the `EntityCompilerPassTest` and adapt it for the `SalesChannelEntityCompilerPass`

**Comment by hlohaus** [comment]
I don't think, i should write unit tests for foreign container config.
If i want to add a missing line in the code.
Additionally they are already a big list of knowing failing tests in php stan.
Maybe its again done, before i check out a new project. :D

I start to writing a compiler pass for the bundle config:

$configParameters = $this->getConfigValues();
        $pluginContainer = new ContainerBuilder();
        $pluginContainer->merge($container);

        $configParameterNames = [];
        foreach ($configParameters as $key => $value) {
            $name = $this->getContainerPrefix() . '.config.' . $key;
            $configParameterNames[] = $name;

            $definition = new Definition();
            $definition->setFactory([new Reference(SystemConfigService::class), 'get']);
            $definition->addArgument($this->getName() . '.config.' . $key);

            $pluginContainer->setParameter($name, $definition);
        }
        $container->merge($pluginContainer);

        $container->addCompilerPass(new EntityRepositoryArgumentCompilerPass());
        $container->addCompilerPass(new RemoveParametersPass($configParameterNames), PassConfig::TYPE_REMOVE);
``
Isn't it a good improvement?

It isn't good idea, to add core function in customer projects.
And i don't like the shopware 5 plugin feeling and
discover new config services in plugins like paypal. :)

@shyim @and-wolf @mitelg Are the coffee still hot?

The sales channel id can come from the request event,
but the services should have a flag for having a no sales channel support.

Today its simply random and a app developer fault?

edit: added more happiness.

**Comment by keulinho** [comment]
You should write unit tests for the container config. With the `EntityCompilerPassTest` there is already an example test case that should be quite easy to adapt here. And additionally the two lines you would add, that register the aliases can also quite good be tested.

Regarding phpstan, there was an error when the DB was not running that in that case many errors popped up, because of a not correctly configured phpstan extension, but this issue was fixed this week in the trunk branch. Besides that phpstan should not throw any errors that are not caused by your changes, as phpstan also runs on every CI run.


Regarding the bundle configuration for extension I'm not quite sure what you want to achieve there, it would be best if you could create a separate PR/issue for that and not mix into this fix.

**Comment by hlohaus** [comment]
I doesn't liked to make more static config to remove other static config.

But i liked your? idea to write a container config. :)

Remember me:
I ithink 2017 the team of best it introduced me in autowiring,
I never heard from  the best plugin developer they (niklas) again.

File: src/Core/System/DependencyInjection/CompilerPass/SalesChannelEntityCompilerPass.php [hunk 1]
```diff
@@ -116,6 +116,8 @@ private function collectDefinitions(ContainerBuilder $container): void
             if (isset($entityNames['fallBack'])) {
                 $repositoryNameMap[$entityNames['fallBack']] = $repositoryId;
             }
+
+            $container->registerAliasForArgument(SalesChannelRepository::class, $entityNames['entityName'] . '.repository', $repositoryId);
```
Comment (keulinho) [inline src/Core/System/DependencyInjection/CompilerPass/SalesChannelEntityCompilerPass.php:120]: To be consistent i would also add a alias for `SalesChannelRepositoryInterface` for each repo

File: tests/unit/php/Core/Framework/Resources/config/test_entity_definition.xml [hunk 2]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" ?>
```
Comment (keulinho) [inline tests/unit/php/Core/Framework/Resources/config/test_entity_definition.xml:1]: this file is never loaded and can probably be removed

File: tests/unit/php/Core/Framework/Resources/config/test_entity_definition_compiled.xml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+        <service id="Shopware\Core\System\SalesChannel\Entity\SalesChannelDefinitionInstanceRegistry">
+            <argument type="string"/>
+            <argument type="string"/>
+            <argument type="string"/>
+            <argument type="string"/>
+        </service>
+
+        <service id="Shopware\Core\System\Country\CountryDefinition"
+                 class="Shopware\Core\System\Country\CountryDefinition">
```
Comment (keulinho) [inline tests/unit/php/Core/Framework/Resources/config/test_entity_definition_compiled.xml:18]: no need to specify the class when it's the same as the `id`, please remove all `class` tags in this file, that will make it shorter and easier to read

Comment (hlohaus) [reply tests/unit/php/Core/Framework/Resources/config/test_entity_definition_compiled.xml:18]: yes, they are dependency, which are not resolved in the compiler class.

File: tests/unit/php/Core/Framework/Resources/config/test_entity_definition.xml (hunk 2 above)
Comment (hlohaus) [reply tests/unit/php/Core/Framework/Resources/config/test_entity_definition.xml:1]: yes, they are dependency, which are not resolved in the compiler class.

File: src/Core/System/DependencyInjection/CompilerPass/SalesChannelEntityCompilerPass.php (hunk 1 above)
Comment (hlohaus) [reply src/Core/System/DependencyInjection/CompilerPass/SalesChannelEntityCompilerPass.php:120]: i test it. it is possible to use this or the other interface,
but if a plugin is not compatible, you can't use both.

It should be fast removed from plugins.

**Comment by hlohaus** [comment]
Hello @keulinho are fired?

File: changelog/_unreleased/2022-07-28-allow-autowiring-of-entity-repositories.md [hunk 4]
```diff
@@ -1,6 +1,7 @@
 ---
-title: Allow autowiring of `EntityRepository` type
+title: Allow autowiring of `EntityRepository` and `SalesChannelRepository` type
 issue: NEXT-22613
 ---
 # Core
 * Changed `\Shopware\Core\Framework\DependencyInjection\CompilerPass\EntityCompilerPass` to create autowiring aliases for the registered entity repositories also for `EntityRepository` type and not just for the deprecated `EntityRepositoryInterface`.
+* Changed `\Shopware\Core\System\DependencyInjection\CompilerPass\SalesChannelEntityCompilerPass` to create autowiring aliases for the registered entity sale channel repositories `SalesChannelRepository`.
```
Comment (shyim) [inline changelog/_unreleased/2022-07-28-allow-autowiring-of-entity-repositories.md:7]: create a new changelog

File: src/Core/System/DependencyInjection/CompilerPass/SalesChannelEntityCompilerPass.php (hunk 1 above)
Comment (shyim) [reply src/Core/System/DependencyInjection/CompilerPass/SalesChannelEntityCompilerPass.php:120]: thats fine that we add only for the new one. So we force the people to use the newer classes

File: tests/unit/php/Core/Framework/DependencyInjection/CompilerPass/SalesChannelEntityCompilerPassTest.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ * @internal
+ * @covers SalesChannelEntityCompilerPass
+ */
+class SalesChannelEntityCompilerPassTest extends TestCase
+{
+    public function testEntityRepositoryAutowiring(): void
+    {
+        $container = new ContainerBuilder();
+
+        $loader = new XmlFileLoader($container, new FileLocator(__DIR__ . '/../../Resources/config'));
```
Comment (shyim) [inline tests/unit/php/Core/Framework/DependencyInjection/CompilerPass/SalesChannelEntityCompilerPassTest.php:21]: Please just `$container->setDefinition` with just plain PHP so it is not depending on any loaders.

Comment (hlohaus) [reply tests/unit/php/Core/Framework/DependencyInjection/CompilerPass/SalesChannelEntityCompilerPassTest.php:21]: No, i explained it jet.

File: changelog/_unreleased/2022-07-28-allow-autowiring-of-entity-repositories.md (hunk 4 above)
Comment (hlohaus) [reply changelog/_unreleased/2022-07-28-allow-autowiring-of-entity-repositories.md:7]: Of course

File: tests/unit/php/Core/Framework/DependencyInjection/CompilerPass/SalesChannelEntityCompilerPassTest.php (hunk 5 above)
Comment (shyim) [reply tests/unit/php/Core/Framework/DependencyInjection/CompilerPass/SalesChannelEntityCompilerPassTest.php:21]: There no reason too use XML only when you are lazy. Just use plain PHP objects

Comment (hlohaus) [reply tests/unit/php/Core/Framework/DependencyInjection/CompilerPass/SalesChannelEntityCompilerPassTest.php:21]: The reason are, we use XML by default. So we can it use by testing.
Else the other problems and fixes will not noticed by us?

**Comment by shyim** [comment]
Hey,

I am closing this PR as it's inactive. Feel free to create a new one when it's updated

**Pull request closed without merging**